chore: string formatter initial commit

devadula-nandan opened the pull request:
Closes #

{{short description}}

#### What did you change?

#### How did you test and verify your work?

devadula-nandan commented on packages/ibm-products-web-components/src/components/string-formatter/string-formatter.ts:
```diff
@@ -0,0 +1,103 @@
+import { property, query } from 'lit/decorators.js';
+import { carbonElement as customElement } from '@carbon/web-components/es/globals/decorators/carbon-element.js';
+import { prefix } from '../../globals/settings';
+// todo: replace with definition tooltip to get tab stop
```
definition tooltip is not ready in carbon/web-components yet.

we could leave this as it is as well.
IMO we don't need a tab stop on string formatter, as it could potentially cause inconsistencies in tab order in some tests. however we could still access it with mouse or touch

elycheea reviewed (commented):
Not a full review, but I know @davidmenendez was considering something like `useOverflowString`, but maybe that’s more for utility format. 

I do think this should be classified as a utility component. But I’d love for us to consider the naming if there are better ways to describe its purpose.

(Maybe @RichKummer has thoughts?)

Also we should label this a `feat` not a `chore` :D

devadula-nandan commented:
> Not a full review, but I know @davidmenendez was considering something like `useOverflowString`, but maybe that’s more for utility format.
> 
> I do think this should be classified as a utility component. But I’d love for us to consider the naming if there are better ways to describe its purpose.
> 
> (Maybe @RichKummer has thoughts?)
> 
> Also we should label this a `feat` not a `chore` :D

ah, the framework agonistic utility, let me try that parallel. sounds like a good idea to me.

RichKummer commented:
`StringFormatter` seems a little generic to me, and could mean anything related to a string. I'm wondering if the name could incorporate `truncate/truncation` and `tooltip` to help convey the purpose? Those are the two key aspects to the utility

`TruncateWithTooltip`? There's also `TruncatedList` so maybe something more in line with that?

devadula-nandan commented:
i have updated this pr, and the changes include.

name change to `truncated string` and the tag name being `cds-truncated-string`.
introduced a new prop `with` which gives two options to present the full string

RichKummer commented:
Hey @devadula-nandan ! 

**Tooltip**
- For this use case, I think `Tooltip` is more appropriate. The `DefinitionToolip` is built "to provide additional information about a particular term or phrase in text content" (from the [Storybook](https://react.carbondesignsystem.com/?path=/docs/components-definitiontooltip--overview)). There's also an impact on the actual tooltip styling (the DefinitionTooltip has a slimmer container.
- That said, the description still be focusable if it's truncated with the tooltip. Does explicitly adding the tabindex have a negative effect on adopting this?

**Expand**
This is looking great, and mirrors the Tearsheet specs. Two things that I'm looking at:
- I can’t tell from the build, but I’m wondering if we could apply the [[Carbon motion package](https://carbondesignsystem.com/elements/motion/code/)](https://carbondesignsystem.com/elements/motion/code/) to provide some motion out of the box. Using the [[Motion generator](https://ibm.github.io/motion/)](https://ibm.github.io/motion/), it looks like Productive motion (UI) using `Easing: cubic-bezier(0.2, 0.2, 0.38, 0.9)` might work for the use case. This could also become a follow-on enhancement.
- For the "read more" indication, it looks like the actual clickable area extends to the left of the "read more" label, including an area with a gradient fade-out. I don't think we necessarily need the fade-out or the expanded clickable area (shown below in the pink border)
<img width="459" alt="Screenshot 2025-05-15 at 11 45 52 AM" src="https://github.com/user-attachments/assets/765dea65-ec28-454e-a4cd-42276fb35307" />

devadula-nandan commented:
> * That said, the description still be focusable if it's truncated with the tooltip. Does explicitly adding the tabindex have a negative effect on adopting this?

it could be an accessibility choice, and we can pass this to accessibility team to get the expected behavior, ill make it tabbable in this pr for now.

> * I can’t tell from the build, but I’m wondering if we could apply the [[Carbon motion package](https://carbondesignsystem.com/elements/motion/code/)](https://carbondesignsystem.com/elements/motion/code/) to provide some motion out of the box. Using the [[Motion generator](https://ibm.github.io/motion/)](https://ibm.github.io/motion/), it looks like Productive motion (UI) using Easing: cubic-bezier(0.2, 0.2, 0.38, 0.9) might work for the use case. This could also become a follow-on enhancement.

i couldn't find the corresponding tokens in the library for the exact vales mentioned, so i have used the nearest available match.
```
// transition: max-height 107ms cubic-bezier(0.2, 0.2, 0.38, 0.9); // recommended
transition: max-height $duration-fast-02 motion(standard, productive); // nearest match for the token
```

> * For the "read more" indication, it looks like the actual clickable area extends to the left of the "read more" label, including an area with a gradient fade-out. I don't think we necessarily need the fade-out or the expanded clickable area (shown below in the pink border)

removed the gradient as per discussion

elycheea commented on packages/ibm-products-web-components/src/components/truncated-text/truncated-text.ts:
```diff
@@ -0,0 +1,214 @@
+
+    return this.with === 'tooltip' && this._isOverflowing
+      ? html`
+          <cds-tooltip align="bottom" enter-delay-ms="0" leave-delay-ms="0">
```
I feel like we should support both `autoAlign` and a manual `tooltipAlign` when using the tooltip.

elycheea commented on packages/ibm-products-web-components/src/components/truncated-text/truncated-text.ts:
```diff
@@ -0,0 +1,214 @@
+  /**
+   * The label on expand button.
+   */
+  @property({ attribute: 'expand-label', type: String, reflect: true })
+  expandLabel = '...more';
+
+  /**
+   * The label on the collapse button.
+   */
+  @property({ attribute: 'collapse-label', type: String, reflect: true })
+  collapseLabel = '...less';
```
I recommend we stick with “View” or “Read” as our defaults?

```suggestion
  @property({ attribute: 'expand-label', type: String, reflect: true })
  expandLabel = 'View more';

  /**
   * The label on the collapse button.
   */
  @property({ attribute: 'collapse-label', type: String, reflect: true })
  collapseLabel = 'View less';
```

elycheea commented on packages/ibm-products-web-components/src/components/truncated-text/truncated-text.scss:
```diff
@@ -0,0 +1,114 @@
+  &.#{$block-class}_button-layered {
+    background-color: $layer;
+  }
+
```
We can try using generated content `::before/::after` to include the ellipses before the expand?

<img width="1360" alt="image" src="https://github.com/user-attachments/assets/90acf09a-fe56-443a-aa66-ad8b7386d7b9" />

<img width="1354" alt="image" src="https://github.com/user-attachments/assets/162f41e6-ff31-48a5-aa50-6f23aa7ea582" />

It’s still included in the click but might be a better in-between since I do think we’d expect to see an ellipsis for truncated text.

elycheea replied:
Found a similar example from Dell Design System although their implementation is a bit more manual — a little akin to the non-linear reading component actually? 🤔 

https://vanilla.delldesignsystem.com/2.46.2/index.html?path=/docs/components-view-more-less--inline

devadula-nandan replied:
👀 looks like a lot of manual setup, also we have to decide at what character we want to truncate and wrap them in spans, which doesn't give the flexibility of having consistent heights for layouts having height constraints, at variable sizes.

in truncated text implementation, when we are truncating by number of lines. i had this challenge of showing the truncation on the text with accounting for `view more` button, which was technically not feasible as webkit line clamp doesn't provide any offset to account for inline button size. a common way design systems tackled this scenario was to add a subtle gradient before the truncation to avoid harsh cut on the text.
we have currently removed it as per the design. but we can add it if we wanted to.
by increasing`0%` in the end to some `20%`
in this line https://github.com/carbon-design-system/ibm-products/pull/7397/files#diff-61d4b66abd22cfa87c99ddcaa80fb43747db3dbd02ea0fecfbb79d5198dffb04R31

devadula-nandan commented on packages/ibm-products-web-components/src/components/truncated-text/truncated-text.mdx:
```diff
@@ -0,0 +1,56 @@
+
+# TruncatedText
+
+The `TruncatedText` component truncates text based on a specified number of
```
agree with changing the wording to utility.

devadula-nandan commented:
the align right seems to be adding some unnecessary styles, it could be adding some rtl direction.

elycheea commented on packages/ibm-products-web-components/src/components/truncated-text/truncated-text.scss:
```diff
@@ -90,6 +87,18 @@ $block-class: #{$prefix}--truncated-text;
+    block-size: 100%;
+    color: $text-primary;
+    content: '...';
+    inline-size: 1em;
+    inset-block-start: 0;
+    inset-inline-start: calc(-1em - 0.2em);
+    padding-inline-start: 0.2em;
```
Any reason we’re using `em` instead of `rem` here? If we’re using these specific numbers over tokens, we should also comment on how we derived these; e.g. avoid magic numbers!

devadula-nandan replied:
i intended `em` particularly in this case as this utility (or component) will be used in various html text elements, that has varying font sizes, so setting `padding-inline-start: 0.2em` would set the tiny padding between `...` and the text to `0.2em` and it scales as per the context (parent font size) its used in. so a tiny text will have tiny padding, and larger text will have larger padding.

elycheea replied:
I actually wonder if `ch` would be better alternative then? But can do that when we move styles to `@carbon/ibm-products-styles` for the React version.

elycheea commented on packages/ibm-products-web-components/src/components/truncated-text/truncated-text.ts:
```diff
@@ -0,0 +1,232 @@
+   * Specify whether a auto align functionality should be applied
+   */
+  @property({ type: Boolean, reflect: true })
+  autoalign = false;
```
Small thing but noticed that React uses `autoAlign` but WC looks like it uses `autoalign`. 👀

devadula-nandan replied:
According to Web Component conventions, properties in the JavaScript context should typically be camelCase (e.g., autoAlign), while their corresponding HTML attributes should be kebab-case (auto-align). In the case of carbon-web-components, the `autoalign` property does not follow this convention. If this is not a deliberate design decision, it might be worth raising a GitHub issue for clarification or consistency.

elycheea reviewed (commented):
A few small comments.

Also wonder if we’Ll need to move some of the styles to `@carbon/ibm-products-styles` once we add the React version.

elycheea reviewed (dismissed):
Actually updating to approve — the comments I have are pretty minor and I think can be considered non-blocking. We can definitely update when we update the React implementation too.

matthewgallo commented on packages/ibm-products-web-components/examples/truncated-text/src/styles.scss:
```diff
@@ -0,0 +1,18 @@
+
+:root {
+  @include theme.theme(themes.$white);
+  background-color: var(--cds-background);
+  color: var(--cds-text-primary);
```
These should be using sass variables

devadula-nandan replied:
looks like all other examples still have them, and coming while cloning a component 👀 , fixed in this.

matthewgallo commented on packages/ibm-products-web-components/src/components/truncated-text/truncated-text.scss:
```diff
@@ -0,0 +1,121 @@
+@use '@carbon/styles/scss/motion' as *;
+
+$prefix: 'c4p';
+$carbon-prefix: 'cds';
```
Can you change this to not hard code the `cds` prefix. It can be access from the styles config

```scss
@use '@carbon/styles/scss/config'

.#{config.$prefix}--some-carbon-component {
  ...
}
```

devadula-nandan replied:
looks like all other examples still have them, and coming while cloning a component 👀 , fixed in this.
